libs/vkd3d-shader/spirv.c

```diff
@@ -11813,6 +11827,56 @@ void vkd3d_shader_extract_feature_meta(struct vkd3d_shader_code *code)
                 meta |= VKD3D_SHADER_META_FLAG_EMITS_TRIANGLES;
 
         }
+        else if ((op == SpvOpDecorate && count == 4) ||
+                (op == SpvOpMemberDecorate && count == 5))
+        {
+            unsigned int delta = op == SpvOpMemberDecorate ? 1 : 0;
+            decoration = spirv[offset + delta + 2];
+
+            if (decoration == SpvDecorationBuiltIn)
+            {
+                builtin = spirv[offset + delta + 3];
+
+                if (builtin == SpvBuiltInSampleMask)
+                {
+                    if (tracked_builtin_count < ARRAY_SIZE(tracked_builtins))
+                    {
+                        struct vkd3d_tracked_builtin *entry = &tracked_builtins[tracked_builtin_count++];
+                        entry->var_id = spirv[offset + 1];
+                        entry->builtin = builtin;
+                    }
+                    else
+                        ERR("Too many tracked built-in variables.\n");
+                }
+            }
+        }
+        else if (op == SpvOpVariable && count >= 4)
+        {
+            storage_class = spirv[offset + 3];
+
+            if (storage_class == SpvStorageClassOutput || storage_class == SpvStorageClassInput)
+            {
+                var_id = spirv[offset + 2];
+
+                for (i = 0; i < tracked_builtin_count; i++)
+                {
+                    const struct vkd3d_tracked_builtin *entry = &tracked_builtins[i];
+
+                    if (entry->var_id != var_id)
```

Comment:
Is it written like this to potentially look for more than builtin later?

Reply:
The idea here was to introduce a mechanism which can be easily extended in the future. I think there's value to that, e.g. if we ever upgrade one of the shader compilers to SPIR-V 1.6 and things like viewport/layer export change, and potential future changes.